app/src/main/cpp/skyline/kernel/scheduler.cpp

```diff
@@ -0,0 +1,367 @@
+
+    void Scheduler::SignalHandler(int signal, siginfo *info, ucontext *ctx, void **tls) {
+        if (*tls) {
+            const auto &state{*reinterpret_cast<nce::ThreadContext *>(*tls)->state};
+            state.scheduler->Rotate(false);
+            YieldPending = false;
+            state.scheduler->WaitSchedule();
+        } else {
+            YieldPending = true;
+        }
+    }
+
+    Scheduler::CoreContext &Scheduler::LoadBalance(const std::shared_ptr<type::KThread> &thread, bool alwaysInsert) {
+        std::lock_guard migrationLock(thread->coreMigrationMutex);
+        auto *currentCore{&cores.at(thread->coreId)};
+
+        if (!currentCore->queue.empty() && thread->affinityMask.count() != 1) {
+            // Select core where the current thread will be scheduled the earliest based off average timeslice durations for resident threads
+            // There's a preference for the current core as migration isn't free
+            size_t minTimeslice{};
+            CoreContext *optimalCore{};
+            for (auto &candidateCore : cores) {
+                if (thread->affinityMask.test(candidateCore.id)) {
+                    u64 timeslice{};
+
+                    if (!candidateCore.queue.empty()) {
+                        std::unique_lock coreLock(candidateCore.mutex);
+
+                        auto threadIterator{candidateCore.queue.cbegin()};
+                        if (threadIterator != candidateCore.queue.cend()) {
+                            const auto &runningThread{*threadIterator};
+                            timeslice += runningThread->averageTimeslice ? std::min(runningThread->averageTimeslice - (util::GetTimeTicks() - runningThread->timesliceStart), 1UL) : runningThread->timesliceStart ? util::GetTimeTicks() - runningThread->timesliceStart : 1UL;
+
+                            while (++threadIterator != candidateCore.queue.cend()) {
+                                const auto &residentThread{*threadIterator};
+                                if (residentThread->priority <= thread->priority)
+                                    timeslice += residentThread->averageTimeslice ? residentThread->averageTimeslice : 1UL;
+                            }
+                        }
+                    }
+
+                    if (!optimalCore || timeslice < minTimeslice || (timeslice == minTimeslice && &candidateCore == currentCore)) {
+                        optimalCore = &candidateCore;
+                        minTimeslice = timeslice;
+                    }
+                }
+            }
+
+            if (optimalCore != currentCore) {
+                if (!alwaysInsert && thread == state.thread)
+                    RemoveThread();
+                else if (!alwaysInsert && thread != state.thread)
+                    [[unlikely]]
+                        throw exception("Migrating an external thread (T{}) without 'alwaysInsert' isn't supported", thread->id);
```

Comment:
It's a hard question as there's no direct equivalent for our load balancing on HOS, it's exclusive to Skyline. The limitation therein is being able to non-cooperatively migrate a thread from one core to another which isn't really done on HOS for any running thread but it can technically do this, we are in the same position of technically being able to support it ever since we moved to per-thread scheduler conditions. 

It's just that it requires additional logic that would waste cycles evaluating in `WaitSchedule` for something that's completely useless, why would a thread need to load balance another running thread?

The reason that the `thread` argument even exists is as it can be used in `KThread::Start` to load balance and insert the thread at the same time, doing it in this function makes things simpler as it locks the migration mutex which would otherwise needed to be handled externally.

Reply:
The last part of writing the previous comment prompted me to take a critical look at the rather poor structure of `LoadBalance` which has now been superseded `GetOptimalCoreForThread` which can be used alongside `InsertThread` or `MigrateToCore` to perform the same function as before. It makes the API far more atomic and neater, the scope of the function was very confusing prior and combining functionality in a way that ended up duplicating code and complicating logic. 

This change also eliminates an issue with usage of `RemoveThread` inside `LoadBalance` which would update `KThread::averageTimeslice` when the thread wasn't scheduled, this isn't done in `MigrateToCore` and this bug which would've otherwise corrupted the heuristic for load balancing has been fixed.

Reply:
cool